File: vol1/ch8/LeetCode506.cpp
struct Player {
    int i;															//运动员编号
    int mark;													//运动员成绩
    Player(int i,int m):i(i),mark(m) {}
    bool operator<(const Player&a) const {
        return mark>a.mark;									//指定按mark递减排列
    }
};
class Solution {
public:
    vector<string> findRelativeRanks(vector<int>& score) {
        int n=score.size();
        set<Player> tset;
        for(int i=0;i<n;i++)
            tset.insert(Player(i,score[i]));
        vector<string> ans(n);
        int i=1;
        for(auto c : tset) {
            if(i==1) ans[c.i]="Gold Medal";
            else if(i==2) ans[c.i]="Silver Medal";
            else if(i==3) ans[c.i]="Bronze Medal";
            else ans[c.i]=to_string(i);
            i++;
        }
        return ans;
    }
};

File: vol1/ch8/LeetCode506.py
from sortedcontainers import SortedList
class Player:
    def __init__(self,i,m):
        self.i=i                #运动员编号
        self.mark=m             #运动员成绩
    
    def __lt__(self,other): 
        if self.mark>other.mark:
            return True					 #指定按mark越大越优先
        else:
            return False

class Solution:
    def findRelativeRanks(self, score: List[int]) -> List[str]:
        n=len(score)
        tset=SortedList()
        for i in range(0,n):
            tset.add(Player(i,score[i]))
        ans=[None]*n
        i=1
        for c in tset:
            if i==1:ans[c.i]="Gold Medal"
            elif i==2:ans[c.i]="Silver Medal"
            elif i==3:ans[c.i]="Bronze Medal"
            else: ans[c.i]=str(i)
            i+=1
        return ans
